hera_sim/cli_utils.py

```diff
@@ -0,0 +1,136 @@
+"""
+import copy
+import os
+import numpy as np
+from .defaults import SEASON_CONFIGS
+from .simulate import Simulator
+
+def get_filing_params(config):
+    """Extract filing parameters from a configuration dictionary."""
+
+    filing_params = dict(
+        outdir=os.getcwd(),
+        outfile_name="hera_sim_simulation.uvh5",
+        output_format="uvh5",
+        clobber=False,
+    )
+    filing_params.update(config.get("filing", {}))
+    if filing_params["output_format"] not in ("miriad", "uvfits", "uvh5"):
+        raise ValueError(
+            "Output format not supported. Please use miriad, uvfits, or uvh5."
+        )
+
+    return filing_params
+
+def validate_config(config):
+    """Validate the contents of a loaded configuration file."""
+    if config.get("defaults", None) is not None:
+        if type(config["defaults"]) is not str:
+            raise ValueError(
+                "Defaults in the CLI may only be specified using a string."
+            )
+        
+        if config["defaults"] in SEASON_CONFIGS.keys():
+            return
+        else:
+            raise ValueError("Default configuration string not recognized.")
+
+    freq_params = config.get("freq", {})
+    time_params = config.get("time", {})
+    array_params = config.get("telescope", {}).get("array_layout", {})
+    if any(param == {} for param in (freq_params, time_params, array_params)):
+        raise ValueError("Insufficient information for initializing simulation.")
+
+    freqs_ok = _validate_freq_params(freq_params)
+    times_ok = _validate_time_params(time_params)
+    array_ok = _validate_array_params(array_params)
+    if not all(freqs_ok, times_ok, array_ok):
+        raise ValueError("Insufficient information to initialize simulation.")
+
+def write_calfits(
+    gains, 
+    filename, 
+    sim=None, 
+    freqs=None, 
+    times=None, 
+    clobber=False
+):
+    """Write gains to disk as a calfits file."""
+    from hera_cal.io import write_cal
+    if sim is not None:
+        if isinstance(sim, Simulator):
+            freqs = sim.freqs * 1e9
+            times = sim.times
+        else:
+            freqs = np.unique(sim.freq_array)
+            times = np.unique(sim.time_array)
+    else:
+        if freqs is None or times is None:
+            raise ValueError(
+                "If a simulation is not provided, then both frequencies and "
+                "times must be specified."
+            )
+
+    # Update gain keys to conform to write_cal assumptions.
+    gains = {(ant, 'x'): gain for ant, gain in gains.items()}
+    write_cal(filename, gains, freqs, times, overwrite=clobber, return_uvc=False)
+
+def write_vis(ref_sim, vis, filename, save_format="uvh5", bda_params=None, **kwargs):
+    """Write a simulated visibility to disk."""
+    tmp_sim = copy.deepcopy(ref_sim)
+    if not isinstance(tmp_sim, Simulator):
+        tmp_sim = Simulator(data=tmp_sim)
+
+    tmp_sim.data.data_array = vis
+    if bda_params:
```

Comment:
I wonder if the bda shouldn't be applied outside this function -- having bda applied inside a function called `write_vis` kind of goes against the "single responsibility principle".

Reply:
I think it probably should be applied outside this function... I'll update it.

Reply:
I ended up just removing this function altogether and put the logic for handling it into the CLI. I think it's better done that way, since there will be some complications with writing stuff to disk when BDA is being applied, both for gain-like and visibility-like terms. I think the implementation that's there now should appropriately handle visibility-like terms--I'm punting the handling of gain-like terms to when I add support for time-varying gains.

Reply:
Cool, sounds good!